Migrate `cip` CLI command to cobra

justaugustus opened the pull request:
#### What type of PR is this?

/kind cleanup feature design

#### What this PR does / why we need it:

We use `cobra` for RelEng tooling, so this PR converts the image promotion to use it as well.

- Use cobra for `cip` command
- Fixup all references to `cip` commands
  - Update cip build target (`//:cip` --> `//cmd/cip:cip`)
  - Use cobra flags
  - Set `//cmd/cip:cip_lib` target visibility to public
  - Update `//:cip-auditor` `container_image` to use `cip audit` subcommand
  - test-e2e/cip: Drop klog flag in command invocation
  - Drop `make test-mac` target
- cip/cmd: Move auditor functionality into separate subcommand
- cip/cmd: Move promotion functionality into separate subcommand
- Update READMEs with new command invocations
- cmd/cip: Minor cleanup of flag descriptions
- cmd/cip: Prefer lowercased output formats (but strings.ToLower() anyway)
- cmd/cip: Add `cip version` command 

#### Which issue(s) this PR fixes:

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.

Fixes #

or

None
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
- Use cobra for `cip` command
- Fixup all references to `cip` commands
  - Update cip build target (`//:cip` --> `//cmd/cip:cip`)
  - Use cobra flags
  - Set `//cmd/cip:cip_lib` target visibility to public
  - Update `//:cip-auditor` `container_image` to use `cip audit` subcommand
  - test-e2e/cip: Drop klog flag in command invocation
  - Drop `make test-mac` target
- cip/cmd: Move auditor functionality into separate subcommand
- cip/cmd: Move promotion functionality into separate subcommand
- cmd/cip: Prefer lowercased output formats (but strings.ToLower() anyway)
- cmd/cip: Add `cip version` command 
```

justaugustus commented:
/test pull-cip-auditor-e2e

justaugustus commented:
/test pull-cip-auditor-e2e

justaugustus commented:
/retest

justaugustus commented:
Ready for review.

/assign @hasheddan @saschagrunert @cpanato @puerco 
cc: @kubernetes-sigs/release-engineering

justaugustus commented:
Flake:
/test pull-cip-auditor-e2e

xmudrii reviewed (commented):
I've quickly gone through the PR, and overall it looks fine to me. I've left two tiny nits about potential improvements.

justaugustus commented:
@xmudrii -- Same thoughts, but I was initially going to take these as follow-ups, as this PR has gotten way larger than I originally wanted/expected.

> Nit: should we have a dedicated `version` command instead of a flag (e.g. like krel)? The dedicated command approach might be a little bit more elegant and easier to implement because we wouldn't have to add logic for checking is the flag provided to each command.

Added in https://github.com/kubernetes-sigs/k8s-container-image-promoter/pull/282/commits/e2965d6f31542acfdd29e8c9cb521fd152cf2835, but I'll take cleanup of the old version commands as a follow-up as they interact with bazel and I'm not sure if modifying them will affect tests yet.

> Tiny nit: why we use uppercase here? Could we make this case-insensitive?

Fixed in https://github.com/kubernetes-sigs/k8s-container-image-promoter/pull/282/commits/b2de725eb98835d89c506565fc2ee495b754f822.

justaugustus commented:
Flake:
/test pull-cip-auditor-e2e

justaugustus commented:
(Successful GCB run: https://console.cloud.google.com/cloud-build/builds/871cfd6d-4a30-4281-97dc-5289ba098a2d?project=k8s-staging-artifact-promoter)

justaugustus commented:
Flake:
/test pull-cip-auditor-e2e

justaugustus commented on cmd/cip/cmd/root.go:
```diff
@@ -0,0 +1,95 @@
+		"test run promotion without modifying any registry",
+	)
+
+	rootCmd.PersistentFlags().BoolVar(
+		&rootOpts.version,
+		"version",
+		rootOpts.version,
+		"print version",
+	)
```
ref: https://github.com/kubernetes-sigs/k8s-container-image-promoter/pull/282#issuecomment-735351685

justaugustus commented on cmd/cip/cmd/run.go:
```diff
@@ -0,0 +1,526 @@
+		case "CSV":
+			snapshot = rii.ToCSV()
+		case "YAML":
+			snapshot = rii.ToYAML(reg.YamlMarshalingOpts{})
```
ref: https://github.com/kubernetes-sigs/k8s-container-image-promoter/pull/282#issuecomment-735351685

saschagrunert reviewed (approved):
/lgtm
